### PR TITLE
feat: Adds the ability to ignore files in directory apps. Closes #2638

### LINKED ISSUE #2638: Add support to ignore certain files in Git source
# Summary

Add support to ignore certain files in Git source.

# Motivation

I have a private Git repository which contains both Kubernetes objects in yaml files and Kafka Connect objects in json files defining Kafka Connect connectors and tasks. Currently when I set up an ArgoCD application pointing at this repo, it will fail to sync because it cannot apply the Kafka Connect objects to the Kubernetes API.

# Proposal

A solution for this could be to allow an extra configuration option in the application `source` configuration to optionally define files, file extensions and / or directories to ignore.

## Comments

I am going to pull down the codebase and see if I can add support for this myself, but I wanted to make you aware of my intention to do this by creating this issue.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Optional. My organization is added to the README.
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

### CODE REVIEW

File: ui/src/app/applications/components/application-parameters/application-parameters.tsx [hunk 1]
```diff
@@ -287,6 +287,11 @@ export const ApplicationParameters = (props: {
             view: (!!directory.recurse).toString(),
             edit: (formApi: FormApi) => <FormField formApi={formApi} field='spec.source.directory.recurse' component={CheckboxField} />
         });
+        attributes.push({
+            title: 'DIRECTORY IGNORE',
+            view: (directory.ignore || []).join(','),
+            edit: (formApi: FormApi) => <FormField formApi={formApi} field='spec.source.directory.ignore' component={ValueArrayInputField} />
```
Comment (alexec) [inline ui/src/app/applications/components/application-parameters/application-parameters.tsx]: does this need help ?

File: pkg/apis/application/v1alpha1/types.go [hunk 2]
```diff
@@ -327,13 +328,49 @@ func (k *ApplicationSourceKsonnet) IsZero() bool {
...
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
 type ApplicationSourceDirectory struct {
 	Recurse bool                     `json:"recurse,omitempty" protobuf:"bytes,1,opt,name=recurse"`
 	Jsonnet ApplicationSourceJsonnet `json:"jsonnet,omitempty" protobuf:"bytes,2,opt,name=jsonnet"`
+	// Ignore is a list of glob patterns to ignore.
```
Comment (jannfis) [inline pkg/apis/application/v1alpha1/types.go]: I think this should be more clear - as in to ignore what and where

**Review by jannfis — COMMENTED** [review]
I like this feature, however I have some concerns about the (behaviour of the ) glob library used. Most people would expect to be able to use extended glob patterns on paths, such as `foo/**/*.yaml` and the `github.com/gobwas/glob` library does support it. However in an unexpected way. 

Consider  these two patterns:
* `/foo/**/*.yaml`
* `/foo/*/*.yaml`

Which of those two patterns would you expect to match against `/foo/bar/baz/test.yaml`? I'd expect the first only to match, but `gobwas/glob` matches for both. 

Quick and easy demonstration vs. `doublestar` library, which behaves as I (and possibly many other people) would expect:

```go
package main

import "fmt"

import "github.com/bmatcuk/doublestar"
import "github.com/gobwas/glob"

func main() {
        p1 := "/foo/**/*.yaml"
        p2 := "/foo/*/*.yaml"
        t1 := "/foo/bar/baz/foo.yaml"
        g1 := glob.MustCompile(p1)
        g2 := glob.MustCompile(p2)
        dm1, _ := doublestar.Match(p1, t1)
        dm2, _ := doublestar.Match(p2, t1)
        gm1 := g1.Match(t1)
        gm2 := g2.Match(t1)
        fmt.Printf("doublestar: %v %v\n", dm1, dm2) 
        fmt.Printf("glob: %v %v\n", gm1, gm2) 
}

```

**Comment by imranismail** [comment]
@alexec why was this closed?

We use mono-repo structure and this feature will greatly improve the sync performance.

**Pull request closed without merging**